### DISCUSSION TITLE
Passing {{ params }} through argparse as a string

### DISCUSSION THREAD

**Discussion by gaborfekete-greehill** [apache/airflow#32710] [Q&A] [answered]
I'd like to pass the whole user defined `{{ params }}` dictionary as an argument to a `KubernetesPodOperator` which runs a script that uses `argparse` to read it and convert it back to a usable `dict`. I'm using `json.dumps` to create the string. I tried creating a task which uses `get_current_context()` then transforms the `params` to a string (without whitespaces).
When using the command and args: `python3 -m my_script --config *params_str*` locally, then it works, using it through the DAG gives me unrecognized argument.

How can I pass the _whole_ user-defined params through?

The code I was using (heavily minimized):

airflow dag:
```
@task
def get_config():
    r = json.dumps(get_current_context()["params"], separators=(",", ":"))
    print(r) # prints r
    print(type(r)) # prints type:str
    print(json.loads(r)) # prints the dict
    print(type(json.loads(r))) # prints type:dict
    return r
    
validation = KubernetesPodOperator(
        cmds=["python3"],
        arguments=[
            "-m",
            "validation",
            "--config",
            get_config(),
        ],)
```
validation.py:
```
parser = ArgumentParser()
parser.add_argument("--config", type=str, required=True)
args, unknown = parser.parse_known_args()
# won't even get here when called from the DAG, works locally
logging.info(args)
logging.info(unknown)
```

**Comment by gaborfekete-greehill** [comment]
I forgot to mention, I'm using Airflow v2.3.0

**Comment by Adaverse** [comment, marked as answer]
Since your task is returning the string, you can use jinja templating to fetch it. 
```
@task
def get_config():
    r = json.dumps(get_current_context()["params"], separators=(",", ":"))
    print(r) # prints r
    print(type(r)) # prints type:str
    print(json.loads(r)) # prints the dict
    print(type(json.loads(r))) # prints type:dict
    return r
    
validation = KubernetesPodOperator(
        cmds=["python3"],
        arguments=[
            "-m",
            "validation",
            "--config",
            "{{ ti.xcom_pull('get_config')['return_value'] }}",
        ],)
```

**Reply by gaborfekete-greehill** [reply]
It returns with the following error:
`jinja2.exceptions.UndefinedError: 'None' has no attribute 'return_value'`

Do I need to enable xcom_push?

**Reply by Adaverse** [reply]
No need to push it explicitly (although you may). Is `get_config` upstream of `KubernetesPodOperator` i.e 
```
get_config() >> validation
```
The above error suggests that get_config does not exist. You have to ensure `get_config` runs before `KubernetesPodOperator`.

**Reply by gaborfekete-greehill** [reply]
it now says this:

`jinja2.exceptions.UndefinedError: 'str object' has no attribute 'return_value'`

should I just remove the `['return_value']` as it is now implicitly a string?

**Reply by gaborfekete-greehill** [reply]
now it gets the string, without the `['return_value']` but it still says that the argument is unrecognized by `argparse`

**Reply by Adaverse** [reply]
Actually I missed your airflow version. In the latest the xcom is pushed into return_value key

**Reply by gaborfekete-greehill** [reply]
I replaced it with `"{{ ti.xcom_pull(task_ids='get_config', key='return_value') }}"` and it seems to work the same, with the unrecognized argument still. I also replaced the return value of `get_config` with a simple `"hello"` string and it still can't recognize it, so I'm afraid that there is something else happening, do I need to convert it to some other type in the airflow before passing it through?
In other operators the `"{{  dag_run.conf['data_preparation']['project_name'] }}"` does work, but I want to pass the whole dictionary, not by each value.

**Comment by Adaverse** [comment]
> now it gets the string, without the `['return_value']` but it still says that the argument is unrecognized by `argparse`

Can you check what is being passed to the argparser. Maybe print the type and value.

**Reply by gaborfekete-greehill** [reply]
it won't even get to the parsing as it fails instantly, I have no variable to print yet.

**Reply by Adaverse** [reply]
Whats the stack trace when it fails?

**Reply by gaborfekete-greehill** [reply]
```
[2023-07-20, 11:55:36 UTC] {pod_manager.py:225} INFO - usage: validation.py [-h]
[2023-07-20, 11:55:36 UTC] {pod_manager.py:225} INFO - validation.py: error: unrecognized arguments: --config hello
[2023-07-20, 11:55:36 UTC] {kubernetes_pod.py:433} INFO - Deleting pod: validation-05d7744540fa408780fa9c6b833a9d5f
[2023-07-20, 11:55:36 UTC] {taskinstance.py:1889} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py", line 402, in execute
    remote_pod=remote_pod,
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py", line 425, in cleanup
    f'Pod {pod and pod.metadata.name} returned a failure:{error_message}\n{remote_pod}'
airflow.exceptions.AirflowException: Pod validation-05d7744540fa408780fa9c6b833a9d5f returned a failure:
```

then it just writes the whole dictionary of the dag, this is the relevant part:

```
'containers': [{'args': [
                     '-m',
                     'validation',
                     '--config',
                     'hello'],
            'command': ['python3'],
```

so it does get the string, but for some reason it won't get sent properly to the script as it can't recognize it

**Reply by Adaverse** [reply]
Just for sake of debugging, can you check using `sys.agrv` on whether it is parsing to confirm on type issue?

**Reply by gaborfekete-greehill** [reply]
I actually just did, and the same problem remains, I cannot pass any argument to my script

**Reply by Adaverse** [reply]
Well again for the sake of debugging can you confirm if the same is true while running it using bash operator instead of k8s pod operator(for confirming if its the problem with the data being sent)

**Reply by gaborfekete-greehill** [reply]
I wrote a `BashOperator` that only echoes the get_config's return value (which was 'hello') and it works
Now I'm trying to output the params

**Reply by gaborfekete-greehill** [reply]
It works with the BashOperator but not with the KubernetesPodOperator, it prints the `params`

**Reply by Adaverse** [reply]
Can you independently try to run k8s pod operator by passing the args and cmd? This will ascertain if its issue with k8s pod operator or k8s client or something to do with k8s not airflow.

**Reply by gaborfekete-greehill** [reply]
I tried using only the `-h` option, and it didn't print any other option, only the help option, no `--config`, nothing

**Comment by gaborfekete-greehill** [comment]
So as it turns out I wasn't using the validation script's latest version, airflow was using one where there wasn't any arguments added to the `argparse` and it was using `parse_args` instead of `parse_known_args`.
@Adaverse's solution to get the `{{ params }}` worked, thank you!

**Reply by Adaverse** [reply]
Sounds about right
